bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/test/java/com/google/cloud/bigtable/mirroring/hbase1_x/TestMirroringTable.java

```diff
@@ -0,0 +1,1411 @@
+    Get get = createGet("g");
+    List<? extends Row> operations = Arrays.asList(increment, append, put, get);
+    mockBatch(
+        primaryTable,
+        increment,
+        createResult("i", "f", "q", 1, "1"),
+        append,
+        createResult("a", "f", "q", 2, "2"),
+        put,
+        new Result(),
+        get,
+        createResult("g", "f", "q", 3, "3"));
+    Object[] results = new Object[operations.size()];
+
+    List<? extends Row> expectedSecondaryOperations =
+        Arrays.asList(
+            createPut("i", "f", "q", 1, "1"),
+            createPut("a", "f", "q", 2, "2"),
+            createPut("p", "f", "q", "v"),
+            createGet("g"));
+    mirroringTable.batch(operations, results);
+    executorServiceRule.waitForExecutor();
+    verify(primaryTable, times(1)).batch(operations, results);
+    ArgumentCaptor<List<? extends Row>> argumentCaptor = ArgumentCaptor.forClass(List.class);
+    verify(secondaryTable, times(1)).batch(argumentCaptor.capture(), any(Object[].class));
+
+    assertPutsAreEqual(
+        (Put) argumentCaptor.getValue().get(0), (Put) expectedSecondaryOperations.get(0));
+    assertPutsAreEqual(
+        (Put) argumentCaptor.getValue().get(1), (Put) expectedSecondaryOperations.get(1));
+    assertPutsAreEqual(
+        (Put) argumentCaptor.getValue().get(2), (Put) expectedSecondaryOperations.get(2));
+    assertThat(argumentCaptor.getValue().get(3)).isEqualTo(expectedSecondaryOperations.get(3));
+  }
+
+  @Test
+  public void testConcurrentWritesAreFlowControlledBeforePrimaryAction()
+      throws IOException, InterruptedException {
+    this.mirroringTable =
+        spy(
+            new MirroringTable(
+                primaryTable,
+                secondaryTable,
+                this.executorServiceRule.executorService,
+                mismatchDetector,
+                flowController,
+                secondaryWriteErrorConsumer,
+                new ReadSampler(100),
+                true,
+                new MirroringTracer()));
+
+    Put put1 = createPut("r1", "f1", "q1", "v1");
+
+    // Both batches should be called even if first one fails.
+    mockBatch(primaryTable, secondaryTable, put1, new IOException());
+
+    InOrder inOrder = Mockito.inOrder(primaryTable, flowController);
+    try {
+      this.mirroringTable.put(put1);
+      fail("should fail");
+    } catch (IOException ignored) {
+    }
+
+    this.executorServiceRule.waitForExecutor();
+
+    inOrder.verify(flowController).asyncRequestResource(any(RequestResourcesDescription.class));
+    inOrder.verify(primaryTable).batch(eq(Arrays.asList(put1)), any(Object[].class));
+
+    verify(primaryTable).batch(eq(Arrays.asList(put1)), any(Object[].class));
+    verify(secondaryTable).batch(eq(Arrays.asList(put1)), any(Object[].class));
+  }
+
+  @Test
+  public void testNonConcurrentOpsWontBePerformedConcurrently()
+      throws IOException, InterruptedException {
+    setupMirroringTableWithDirectExecutor();
```

Comment:
The name shoudl tell that it creates a concurrent write table?

Reply:
Done.